apps/andi/lib/andi/dataset_cache.ex

```diff
@@ -4,26 +4,46 @@ defmodule Andi.DatasetCache do
 
   require Logger
 
-  alias Andi.Services.DatasetRetrieval
+  import Andi, only: [instance_name: 0]
 
-  # Client
   def start_link(_) do
     GenServer.start_link(__MODULE__, [], name: __MODULE__)
   end
 
-  def put_datasets(datasets) when is_list(datasets) do
-    Enum.each(datasets, &put_dataset/1)
+  def put(datasets) when is_list(datasets) do
+    Enum.each(datasets, &put/1)
   end
 
-  def put_dataset(%SmartCity.Dataset{} = dataset) do
-    :ets.insert(__MODULE__, {dataset.id, dataset})
+  def put(%SmartCity.Dataset{} = dataset) do
+    updated =
+      dataset.id
+      |> get()
+      |> Map.merge(%{"id" => dataset.id, "dataset" => dataset})
+
+    :ets.insert(__MODULE__, {dataset.id, updated})
+  end
+
+  def put(%{"id" => id, "ingested_time" => time_stamp}) do
+    updated =
+      id
+      |> get()
+      |> Map.merge(%{"id" => id, "ingested_time" => time_stamp})
+
+    :ets.insert(__MODULE__, {id, updated})
   end
 
-  def put_dataset(invalid_dataset) do
+  def put(invalid_dataset) do
     Logger.warn("Not caching dataset because it is invalid: #{inspect(invalid_dataset)}")
   end
 
-  def get_datasets do
+  defp get(id) do
+    case :ets.match(__MODULE__, {id, :"$1"}) do
+      [[h | _inner_t] | _t] -> h
```

Comment:
What is this line of code doing?

Reply:
Also if you're trying to get just one thing, you might look at `:ets.match_object`. It should return a list of the tuples you put into the table.

Reply:
Good catch; we've updated it.